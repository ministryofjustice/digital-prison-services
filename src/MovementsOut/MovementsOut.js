import React from 'react'
import PropTypes from 'prop-types'
import GridRow from '@govuk-react/grid-row'
import GridCol from '@govuk-react/grid-col'
import Flags from '../FullFlags/Flags'
import SortableColumn from '../tablesorting/SortableColumn'
import OffenderName from '../OffenderName'
import DateFormatter from '../DateFormatter'
import OffenderImage from '../OffenderImage'
import SortLov from '../tablesorting/SortLov'
import HoursAndMinutes from '../HoursAndMinutes'
import { LAST_NAME } from '../tablesorting/sortColumns'

const MovementsOut = ({ rows, sortOrder, setColumnSort }) => (
  <>
    <GridRow>
      <GridCol setWidth="one-quarter">
        <SortLov sortColumns={[LAST_NAME]} sortColumn={LAST_NAME} sortOrder={sortOrder} setColumnSort={setColumnSort} />
      </GridCol>
    </GridRow>
    <table>
      <thead>
        <tr>
          {/* eslint-disable-next-line jsx-a11y/control-has-associated-label */}
          <th className="straight width5" />
          <th className="straight width15">
            <SortableColumn
              heading="Name"
              column={LAST_NAME}
              sortOrder={sortOrder}
              setColumnSort={setColumnSort}
              sortColumn={LAST_NAME}
            />
          </th>
          <th className="straight width5">Prison no.</th>
          <th className="straight width5">D.O.B.</th>
          <th className="straight width5">Time out</th>
          <th className="straight width10">Reason</th>
          <th className="straight width15">Flags</th>
        </tr>
      </thead>
      <tbody>
        {rows.map(row => (
          <tr className="row-gutters" key={[row.offenderNo, row.reasonDescription, row.timeOut].join('_')}>
            <td className="row-gutters">
              <OffenderImage offenderNo={row.offenderNo} />
            </td>
            <td className="row-gutters">
              <OffenderName lastName={row.lastName} firstName={row.firstName} />
            </td>
            <td className="row-gutters">{row.offenderNo}</td>
            <td className="row-gutters">
              <DateFormatter isoDate={row.dateOfBirth} />
            </td>
            <td>
              <HoursAndMinutes hhmmss={row.timeOut} />
            </td>
            <td>{row.reasonDescription}</td>
            <td className="row-gutters">
              <Flags alerts={row.alerts} />
            </td>
          </tr>
        ))}
      </tbody>
    </table>
  </>
)

MovementsOut.propTypes = {
  rows: PropTypes.arrayOf(
    PropTypes.shape({
      offenderNo: PropTypes.string.isRequired,
      firstName: PropTypes.string.isRequired,
      lastName: PropTypes.string.isRequired,
      timeOut: PropTypes.string,
      reasonDescription: PropTypes.string,
      alerts: PropTypes.arrayOf(PropTypes.string).isRequired,
    })
  ),
  sortOrder: PropTypes.string,
  setColumnSort: PropTypes.func,
}
MovementsOut.defaultProps = {
  rows: [],
  sortOrder: '',
  setColumnSort: () => {},
}

export default MovementsOut
