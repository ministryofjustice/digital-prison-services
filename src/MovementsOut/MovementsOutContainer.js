import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'

import axios from 'axios'

import MovementsOut from './MovementsOut'
import { resetError, setLoaded } from '../redux/actions'
import { ASC } from '../tablesorting/sortOrder'
import Page from '../Components/Page'
import routePaths from '../routePaths'

import SortableDataSource from '../tablesorting/SortableDataSource'
import { lastNameComparator } from '../tablesorting/comparatorComposition'

class MovementsOutContainer extends Component {
  constructor(props) {
    super(props)
    this.state = {
      movementsOut: [],
    }
  }

  componentDidMount() {
    this.getMovementsOut()
  }

  componentDidUpdate(previousProps) {
    const { agencyId, history } = this.props

    if (agencyId !== previousProps.agencyId) {
      history.push(routePaths.establishmentRoll)
    }
  }

  async getMovementsOut() {
    const { agencyId, handleError, resetErrorDispatch, setLoadedDispatch } = this.props
    try {
      resetErrorDispatch()
      setLoadedDispatch(false)
      const response = await axios.get(`/api/movements/${agencyId}/out`)
      const movementsOut = response.data
      this.setState({ movementsOut })
      setLoadedDispatch(true)
    } catch (error) {
      handleError(error)
    }
  }

  render() {
    const { movementsOut } = this.state

    return (
      <Page title="Out today">
        <SortableDataSource sortOrder={ASC} rows={movementsOut} comparator={lastNameComparator}>
          <MovementsOut />
        </SortableDataSource>
      </Page>
    )
  }
}

MovementsOutContainer.propTypes = {
  handleError: PropTypes.func.isRequired,
  agencyId: PropTypes.string.isRequired,
  resetErrorDispatch: PropTypes.func.isRequired,
  setLoadedDispatch: PropTypes.func.isRequired,
  history: PropTypes.shape({
    replace: PropTypes.func.isRequired,
    push: PropTypes.func.isRequired,
  }).isRequired,
}

const mapStateToProps = state => ({
  agencyId: state.app.user.activeCaseLoadId,
})

const mapDispatchToProps = dispatch => ({
  resetErrorDispatch: () => dispatch(resetError()),
  setLoadedDispatch: status => dispatch(setLoaded(status)),
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(MovementsOutContainer)
