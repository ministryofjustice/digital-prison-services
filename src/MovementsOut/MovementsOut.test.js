import React from 'react'
import { shallow } from 'enzyme'

import MovementsOut from './MovementsOut'

describe('Movements out', () => {
  it('should render page correctly', () => {
    const movements = [
      {
        offenderNo: 'offenderNo',
        firstName: 'firstName',
        lastName: 'lastName',
        timeOut: '12:00',
        reasonDescription: 'reasonDescription',
        alerts: ['HA', 'XSA', 'XA', 'PEEP', 'XEL', 'XRF', 'XTACT'],
      },
    ]
    const wrapper = shallow(<MovementsOut sortOrder="ASC" setColumnSort={jest.fn()} movementsOut={movements} />)

    expect(wrapper).toMatchSnapshot()
  })
})
