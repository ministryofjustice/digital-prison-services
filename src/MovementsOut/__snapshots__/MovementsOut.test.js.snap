// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movements out should render page correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MovementsOut
    movementsOut={
      Array [
        Object {
          "alerts": Array [
            "HA",
            "XSA",
            "XA",
            "PEEP",
            "XEL",
            "XRF",
            "XTACT",
          ],
          "firstName": "firstName",
          "lastName": "lastName",
          "offenderNo": "offenderNo",
          "reasonDescription": "reasonDescription",
          "timeOut": "12:00",
        },
      ]
    }
    setColumnSort={[MockFunction]}
    sortOrder="ASC"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <SortLov
          setColumnSort={[MockFunction]}
          sortColumn="lastName"
          sortColumns={
            Array [
              "lastName",
            ]
          }
          sortOrder="ASC"
        />,
        <MovementsOutTable
          movementsOut={
            Array [
              Object {
                "alerts": Array [
                  "HA",
                  "XSA",
                  "XA",
                  "PEEP",
                  "XEL",
                  "XRF",
                  "XTACT",
                ],
                "firstName": "firstName",
                "lastName": "lastName",
                "offenderNo": "offenderNo",
                "reasonDescription": "reasonDescription",
                "timeOut": "12:00",
              },
            ]
          }
          setColumnSort={[MockFunction]}
          sortOrder="ASC"
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "setColumnSort": [MockFunction],
          "sortColumn": "lastName",
          "sortColumns": Array [
            "lastName",
          ],
          "sortOrder": "ASC",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "movementsOut": Array [
            Object {
              "alerts": Array [
                "HA",
                "XSA",
                "XA",
                "PEEP",
                "XEL",
                "XRF",
                "XTACT",
              ],
              "firstName": "firstName",
              "lastName": "lastName",
              "offenderNo": "offenderNo",
              "reasonDescription": "reasonDescription",
              "timeOut": "12:00",
            },
          ],
          "setColumnSort": [MockFunction],
          "sortOrder": "ASC",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <SortLov
            setColumnSort={[MockFunction]}
            sortColumn="lastName"
            sortColumns={
              Array [
                "lastName",
              ]
            }
            sortOrder="ASC"
          />,
          <MovementsOutTable
            movementsOut={
              Array [
                Object {
                  "alerts": Array [
                    "HA",
                    "XSA",
                    "XA",
                    "PEEP",
                    "XEL",
                    "XRF",
                    "XTACT",
                  ],
                  "firstName": "firstName",
                  "lastName": "lastName",
                  "offenderNo": "offenderNo",
                  "reasonDescription": "reasonDescription",
                  "timeOut": "12:00",
                },
              ]
            }
            setColumnSort={[MockFunction]}
            sortOrder="ASC"
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "setColumnSort": [MockFunction],
            "sortColumn": "lastName",
            "sortColumns": Array [
              "lastName",
            ],
            "sortOrder": "ASC",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "movementsOut": Array [
              Object {
                "alerts": Array [
                  "HA",
                  "XSA",
                  "XA",
                  "PEEP",
                  "XEL",
                  "XRF",
                  "XTACT",
                ],
                "firstName": "firstName",
                "lastName": "lastName",
                "offenderNo": "offenderNo",
                "reasonDescription": "reasonDescription",
                "timeOut": "12:00",
              },
            ],
            "setColumnSort": [MockFunction],
            "sortOrder": "ASC",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
